### ISSUE TITLE
Clearing a session's data does not get persisted to the store

### ISSUE THREAD

**Issue by Hirevo** [maxcountryman/tower-sessions#21]
I encountered an issue in `0.2.0` where removing a value from the session would not be considered a change and therefore not saved back into the store.  

This problem only shows when the removed value was the very last one in the session's data, leaving it empty.

Here is a minimal example of the problem:

```rust
use tower_sessions::Session;

fn main() {
    let session = Session::default();
    assert!(!session.modified());

    session.insert("foo", 42);
    assert!(session.modified());
    //      ^^^^^^^^^^^^^^^^^^^ returns `true`, as expected

    session.remove_value("foo");
    assert!(session.modified());
    //      ^^^^^^^^^^^^^^^^^^^ suddenly returns `false`, despite the internal `modified` flag being set
}
```

Other methods like `session.clear()` and `session.flush()` can trigger this issue.

My use case was that I was removing a `user_id` field from the session to implement a logout endpoint.  
This issue caused the logout button to effectively do nothing, since the removal wasn't persisted anymore.

**Comment by Hirevo** [comment]
This issue seems to be caused by the addition of an `inner.data.is_empty()` check in `Session::modified`:

https://github.com/maxcountryman/tower-sessions/blob/5a1f7014c0acea38c827eb693f93e05ef09c0ffb/src/session.rs#L452-L455

**Comment by maxcountryman** [comment]
In this case the session is not considered modified. However, it should be deleted. (Perhaps it's not being deleted?)

Semantically, an empty session should never be saved to the store. This mirrors what Django implementation and is also important for negative caching.

**Comment by maxcountryman** [comment]
I think it's a bug that remove doesn't account for this: if the session ends up empty, we need to delete the session. Specifically this is a problem if we've saved the session to the store previously, since it won't be updated again (i.e. deleted).

**Comment by Hirevo** [comment]
After some more thought, I agree that if the session would end up empty, the session should likely auto-delete itself.  

I've tried doing the following to fix this manually, like this:

```diff
  session.remove_value("author.id");
+ if session.is_empty() {
+     session.delete();
+ }
```

But this did not work, due to `is_empty` always returning `false` if `expiration_time` is set, and there is currently no way of unsetting an `expiration_time`.  

The only way would be to convert it to a `SessionRecord` and check `record.data.is_empty()`, like so.

```diff
  session.remove_value("author.id");
- if session.is_empty() {
+ let record = SessionRecord::from(&session);
+ if record.data().is_empty() {
    session.delete();
}
```

This does work and mitigates the issue, but this clones the session's data twice (in the `From` impl and when calling `SessionRecord::data`), which isn't ideal.

**Comment by maxcountryman** [comment]
I think something like this might work for the the time being:

```diff
diff --git a/src/service.rs b/src/service.rs
index bb4c1c8..12684aa 100644
--- a/src/service.rs
+++ b/src/service.rs
@@ -118,6 +118,13 @@ where
                 }
             };

+            // In order to ensure removing the last value of a session updates the store, we check for an empty session. Empty sessions should be removed from the store.
+            if session.is_empty() {
+                session_store.delete(&session.id()).await?;
+                cookies.remove(cookie_config.build_cookie(&session));
+                return res;
+            }
+
             // For further consideration:
             //
             // We only persist the session in the store when the `modified` flag is set.
diff --git a/src/session.rs b/src/session.rs
index f16978a..aa4339a 100644
--- a/src/session.rs
+++ b/src/session.rs
@@ -495,7 +495,7 @@ impl Session {
     /// ```
     pub fn is_empty(&self) -> bool {
         let inner = self.inner.lock();
-        inner.expiration_time.is_none() && inner.data.is_empty()
+        inner.data.is_empty()
     }
 }
```

**Comment by maxcountryman** [comment]
It might be better to move to a state enum like actix uses for their session implementation in the future.

Another downside here is that negative caching is effectively broken, but I think that's inevitable. (It probably requires a larger API change and it's unclear how valuable that really is, especially if the removal cookie is being set.)

Anyway, if you agree this will work for your use case I can update accordingly.

**Comment by maxcountryman** [comment]
I've published `v0.2.1` to address. Thank you for reporting this!
